File: java/main/java/leetcode/Subsets.java
package leetcode;

import org.junit.Assert;
import org.junit.Test;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * https://leetcode.com/problems/subsets/
 * Created by jeremy on 2019/11/11.
 */
public class Subsets {
    public static List<List<Integer>> subsets(int[] nums) {
        return subsets(nums, new ArrayList<>(), 0);
    }

    public static List<List<Integer>> subsets(int[] nums, List<Integer> candidate, int index) {
        if (index >= nums.length) {
            return Collections.singletonList(candidate);
        } else {
            List<Integer> nextCandidate = Stream.of(candidate, Collections.singletonList(nums[index]))
                .flatMap(Collection::stream)
                .collect(Collectors.toList());
            return Stream.of(subsets(nums, candidate, index + 1), subsets(nums, nextCandidate, index + 1))
                .flatMap(Collection::stream)
                .collect(Collectors.toList());
        }
    }

    public static class UnitTest {
        @Test
        public void test0() {
            List<List<Integer>> expected = Collections.singletonList(Collections.emptyList());;
            Assert.assertEquals(expected, subsets(new int[] {}));
        }

        @Test
        public void test1() {
            List<List<Integer>> expected = new ArrayList<>();
            expected.add(Collections.emptyList());
            expected.add(Collections.singletonList(3));
            expected.add(Collections.singletonList(2));
            expected.add(Stream.of(2, 3).collect(Collectors.toList()));
            expected.add(Collections.singletonList(1));
            expected.add(Stream.of(1, 3).collect(Collectors.toList()));
            expected.add(Stream.of(1, 2).collect(Collectors.toList()));
            expected.add(Stream.of(1, 2, 3).collect(Collectors.toList()));

            Assert.assertEquals(expected, subsets(new int[] { 1, 2, 3 }));
        }
    }
}

File: kotlin/main/kotlin/com/sungjk/acmicpc/leetcode/Subsets.kt
package com.sungjk.acmicpc.leetcode

import org.junit.Assert
import org.junit.Test

/**
 * Created by jeremy on 2019/10/15.
 */
class Subsets {

    fun subsets(nums: IntArray): List<List<Int>> {
        if (nums.isEmpty()) return listOf(emptyList())
        this.subsets(nums, Math.pow(2.0, nums.size.toDouble()).toInt(), 0, mutableListOf())
        return listOf(emptyList())
    }
    // 2 [1]: [], [1]
    // 4 [1, 2]: [], [1], [2], [1, 2]
    // 8: [1, 2, 3]: [], [1], [2], [3], [1, 2], [1, 3], [2, 3], [1, 2, 3]
    // 16: [1, 2, 3, 4]: ...
    fun subsets(nums: IntArray, end: Int, start: Int, subsets: List<List<Int>>) {

    }

    @Test
    fun test0() {
        Assert.assertEquals(listOf(emptyList<Int>()), subsets(intArrayOf()))
    }

    @Test
    fun test1() {
        val expected = listOf(
            listOf(3),
            listOf(1),
            listOf(2),
            listOf(1,2,3),
            listOf(1,3),
            listOf(2,3),
            listOf(1,2)
        )
        Assert.assertEquals(expected, subsets(intArrayOf(1, 2, 3)))
    }

}
